internal/fwschemadata/data_default.go

```diff
@@ -75,76 +75,229 @@ func (d *Data) TransformDefaults(ctx context.Context, configRaw tftypes.Value) d
+			}
+
+			logging.FrameworkTrace(ctx, fmt.Sprintf("setting attribute %s to default value: %s", fwPath, resp.PlanValue))
+
+			return resp.PlanValue.ToTerraformValue(ctx)
 		case fwschema.AttributeWithMapDefaultValue:
 			defaultValue := a.MapDefaultValue()
-			if defaultValue != nil {
-				resp := defaults.MapResponse{}
-				defaultValue.DefaultMap(ctx, defaults.MapRequest{}, &resp)
-				logging.FrameworkTrace(ctx, fmt.Sprintf("setting attribute %s to default value: %s", fwPath.String(), resp.PlanValue.String()))
-				return resp.PlanValue.ToTerraformValue(ctx)
+
+			if defaultValue == nil {
+				return tfTypeValue, nil
 			}
+			req := defaults.MapRequest{
+				Path: fwPath,
+			}
+			resp := defaults.MapResponse{}
+
+			defaultValue.DefaultMap(ctx, req, &resp)
+
+			diags.Append(resp.Diagnostics...)
+
+			if resp.Diagnostics.HasError() {
+				return tfTypeValue, nil
+			}
+
+			if resp.PlanValue.ElementType(ctx) == nil {
+				logging.FrameworkWarn(ctx, "attribute default declared, but returned no value")
+
+				return tfTypeValue, nil
+			}
+
+			logging.FrameworkTrace(ctx, fmt.Sprintf("setting attribute %s to default value: %s", fwPath, resp.PlanValue))
+
+			return resp.PlanValue.ToTerraformValue(ctx)
 		case fwschema.AttributeWithNumberDefaultValue:
 			defaultValue := a.NumberDefaultValue()
-			if defaultValue != nil {
-				resp := defaults.NumberResponse{}
-				defaultValue.DefaultNumber(ctx, defaults.NumberRequest{}, &resp)
-				logging.FrameworkTrace(ctx, fmt.Sprintf("setting attribute %s to default value: %s", fwPath.String(), resp.PlanValue.String()))
-				return resp.PlanValue.ToTerraformValue(ctx)
+
+			if defaultValue == nil {
+				return tfTypeValue, nil
+			}
+
+			req := defaults.NumberRequest{
+				Path: fwPath,
+			}
+			resp := defaults.NumberResponse{}
+
+			defaultValue.DefaultNumber(ctx, req, &resp)
+
+			diags.Append(resp.Diagnostics...)
+
+			if resp.Diagnostics.HasError() {
+				return tfTypeValue, nil
 			}
+
+			logging.FrameworkTrace(ctx, fmt.Sprintf("setting attribute %s to default value: %s", fwPath, resp.PlanValue))
+
+			return resp.PlanValue.ToTerraformValue(ctx)
 		case fwschema.AttributeWithObjectDefaultValue:
 			defaultValue := a.ObjectDefaultValue()
-			if defaultValue != nil {
-				resp := defaults.ObjectResponse{}
-				defaultValue.DefaultObject(ctx, defaults.ObjectRequest{}, &resp)
-				logging.FrameworkTrace(ctx, fmt.Sprintf("setting attribute %s to default value: %s", fwPath.String(), resp.PlanValue.String()))
-				return resp.PlanValue.ToTerraformValue(ctx)
+
+			if defaultValue == nil {
+				return tfTypeValue, nil
+			}
+
+			req := defaults.ObjectRequest{
+				Path: fwPath,
 			}
+			resp := defaults.ObjectResponse{}
+
+			defaultValue.DefaultObject(ctx, req, &resp)
+
+			diags.Append(resp.Diagnostics...)
+
+			if resp.Diagnostics.HasError() {
+				return tfTypeValue, nil
+			}
+
+			logging.FrameworkTrace(ctx, fmt.Sprintf("setting attribute %s to default value: %s", fwPath, resp.PlanValue))
```

Comment:
curious q: Is there a reason why collection types need a warning nil check and the structural type, Object, doesn't?

Referring to the other `resp.PlanValue.ElementType(ctx) == nil` checks for map, set, list

Reply:
We could consider an object value check here to ensure that a provider defined object value's type matches the schema's defined/expected type and raise a warning/error in that case, rather than having it error/panic later in a different way.

Reply:
I created a quick note for myself to double check the framework's current behavior in the above mentioned scenario and will raise an issue if it feels warranted to adjust that behavior.